### PR TITLE
Prefer phoenix router start events over Plug.Telemetry events

### PR DESCRIPTION
**Pull request by d-webs**
See https://github.com/phoenixframework/phoenix/issues/3927

The`[:phoenix, :endpoint, :start]` event is not guaranteed to
be executed before `[:phoenix, :router_dispatch, :start]`, and in most
cases it is not. This changes the event handlers to rely only on
`[:phoenix, :router_dispatch, :start]` rather than `[:phoenix,
:endpoint, :start]`

### CODE REVIEW

File: lib/opentelemetry_phoenix.ex [hunk 1]
```diff
@@ -97,11 +75,12 @@ defmodule OpentelemetryPhoenix do
   end
 
   @doc false
-  def handle_endpoint_start(_event, _measurements, %{conn: %{adapter: adapter} = conn}, _config) do
+  def handle_router_dispatch_start(_event, _measurements, %{conn: conn} = meta, _config) do
     # TODO: maybe add config for what paths are traced? Via sampler?
     :otel_propagator.text_map_extract(conn.req_headers)
 
-    span_name = "HTTP #{conn.method}"
+    span_name = "#{conn.method} #{meta.route}"
```
Comment (fujin) [inline lib/opentelemetry_phoenix.ex:93]: Is `meta.route` here sanitized or anything, e.g. placeholders instead of real values?

Comment (d-webs) [reply lib/opentelemetry_phoenix.ex:93]: yep it would return `/v1/merchants/:id`, for example

**Comment by bryannaegele** [comment]
Can you explain what you mean by the endpoint start event not being guaranteed to execute to execute before the router start event? Unless the telemetry plug was inadvertently put into the router file, this is not the case unless the user was upgrading and omitted the addition of the plug.

I am very hesitant to make this change as this in turn would require users to add yet another package to begin their traces higher in the callstack. Until there is an integration and bridge library at the cowboy handler level, this change would result in a worse outcome for what most users would expect to get out of this package.

Are you interested in tackling the cowboy bridge @dwebster17? https://github.com/beam-telemetry/cowboy_telemetry

**Comment by d-webs** [comment]
@bryannaegele thanks for the clarification. I quickly realized that this was absolutely a misunderstanding on my end :) I'm going to go ahead and close this, and happy to look into the cowboy bridge!

**Comment by bryannaegele** [comment]
I was thinking about this some more this morning and a couple of other issues this proposed change would raise which need to be addressed, one of which I do not think can be done by removing reliance on Plug.Telemetry being present in the endpoint. We've had a _lot_ of discussions on how to remove reliance on it, all of which require depending on doing some things in Cowboy.

1. We can only parse and set the parent ID from the trace ctx headers once and it has to be done at the top-level span. This change would produce an incorrect order of spans if users did that higher in the chain. Once a cowboy bridge is created, we'd need to add a check that we're not already in a trace before parsing that header and setting the parent.

2. The original design was to produce a single span for a request within Phoenix and enhancing or modifying that span as it went through further events. Spans aren't free, so my feeling is we should as a guiding principle try to keep them to one if at all possible. I believe we could make better use of events to mark timings, such as hitting the router, while keeping with the one span.

3. This change would prevent capturing any request that didn't make it to the router which is not what I would expect. It also cuts out potentially large amounts of time off the request when plugs are added before the router is hit. For me, this is a blocker to this proposal.

I understand Jóse's comment that you can't rely on the plug being in the correct place or there at all, but that wasn't meant to suggest that we can't say you have to have it for this library. That's our choice as library authors and we know that having it is in fact necessary to give them the best experience possible with the least amount of work. If the primary concern then is that we require it, then let's add it to the docs that it is required and provide guidance on where to put it, what the tradeoffs are on where you place it, etc.

**Comment by bryannaegele** [comment]
@dwebster17 no worries. Let me know your thoughts on the points I just posted, as well. I think there's a number of things we can do here to improve things, just that removing reliance on the plug being present doesn't accomplish the goals I have for this lib.

**Pull request closed without merging**